### ISSUE TITLE
Custom Cache Repository for RestartCommand and WorkCommand

### ISSUE THREAD

**Issue by infinitum11** [laravel/ideas#2401]
### Problem's Description

In order to take advantage of gracefully [worker restart feature](https://laravel.com/docs/8.x/queues#queue-workers-and-deployment) we MUST set the **default cache store** (in config file) that supports "locking". The default file cache doesn't support it. I tried several ways to change cache repository of "WorkCommand" for but without luck.

1. Firstly, I tried to take advantage of "Contextual Binding", but it has not effect because "WorkCommand" class is not initialized via container.

app\Providers\AppServiceProvider.php

```php

$this->app
    ->when("command.queue.work")
    ->needs('cache.store')
    ->give(function () {
        return Cache::store('file');
    });
```
src\Illuminate\Foundation\Providers\ArtisanServiceProvider.php

```php
$this->app->singleton('command.queue.work', function ($app) {
    return new QueueWorkCommand($app['queue.worker'], $app['cache.store']);
});
```

2. Next, I tried to use Container Events to subscribe to event when "WorkCommand" is resolved and change cache component. No luck here because **$cache** property is **protected** and there is no cache setter.

app\Providers\AppServiceProvider.php
```php
$this->app->resolving('command.queue.work', function ($command, $app) {
    $command->cache = Cache::store('file'); // Error! The "cache" property is protected
});
```

src\Illuminate\Queue\Console\WorkCommand.php
```php
/**
 * The cache store implementation.
 *
 * @var \Illuminate\Contracts\Cache\Repository
 */
protected $cache;
```

3. Finally, I tried to override container's "command.queue.work" instance, but it also doesn't work because framework's service providers are initialized before app's service providers.

app\Providers\AppServiceProvider.php
```php
$this->app->singleton('command.queue.work', function ($app) {
    return new WorkCommand($app['queue.worker'], Cache::store('database'));
});
```

src\Illuminate\Foundation\Application.php
```php
    public function registerConfiguredProviders()
    {
        $providers = Collection::make($this->config['app.providers'])
                        ->partition(function ($provider) {
                            return strpos($provider, 'Illuminate\\') === 0;
                        });

        $providers->splice(1, 0, [$this->make(PackageManifest::class)->providers()]);

        (new ProviderRepository($this, new Filesystem, $this->getCachedServicesPath()))
                    ->load($providers->collapse()->toArray());
    }
```

### Solution

For now I see only one solution - extend both classes and override cache component in a constructor, but I believe this is a bad practice.

Can we still have file cache as a default cache storage, but also have possibility to configure cache storage for RestartCommand and WorkCommand?

**Comment by paras-malhotra** [comment]
Did you try this?

```php
$this->app
    ->when('command.queue.work')
    ->needs('cache.store')
    ->give(function () {
        return Cache::store('database');
    });
```

**Comment by infinitum11** [comment]
It doesn't work either.

**Comment by paras-malhotra** [comment]
Yeah, here's the issue:

1. The `ArtisanServiceProvider` is registered after the `AppServiceProvider` because it's a deferrable service provider. So, you cannot rebind another instance or singleton.
2. `cache.store` is a singleton. So, it is not ever "built" again by the `Container`. This means that contextual bindings won't work because the container doesn't re-build singletons, it just picks it up from its cached array.

You have two options:
1. Extend the `WorkCommand` class and register a new command
2. Extend the `ConsoleSupportServiceProvider` class to provide your own `ArtisanServiceProvider` that registers the queue work commands in a different manner

**Comment by paras-malhotra** [comment]
@themsaid currently, there is no way to re-bind deferred services because only one provider can "provide" a deferred service. Is that a correct understanding? Seems to me that there may be many use cases where users would want to re-bind deferred services.

**Comment by infinitum11** [comment]
> 1. Extend the WorkCommand class and register a new command
> 2. Extend the ConsoleSupportServiceProvider class to provide your own ArtisanServiceProvider that registers the queue work commands in a different manner

As I understand these workarounds will work until the constructor's signature of framework classes changes.

**Comment by paras-malhotra** [comment]
> @paras-malhotra, as I understand these workarounds will work until the constructor's signature of framework classes changes. Am I right?

That's correct, yes. The constructors will not change in minor versions, as that is considered as a breaking change. So, I don't think its something you need to be concerned about.
